packages/core/integration-tests/test/cache.js

```diff
@@ -0,0 +1,370 @@
+import assert from 'assert';
+import path from 'path';
+import {bundle, run, overlayFS, ncp} from '@parcel/test-utils';
+
+function runBundle() {
+  return bundle(path.join(__dirname, '/input/src/index.js'), {
+    inputFS: overlayFS,
+    disableCache: false,
+  });
+}
+
+async function testCache(update) {
+  // Delete cache from previous test and perform initial build
+  await overlayFS.rimraf(path.join(__dirname, '/input'));
+  await ncp(
+    path.join(__dirname, '/integration/cache'),
+    path.join(__dirname, '/input'),
+  );
+  await overlayFS.rimraf(path.join(__dirname, '/input/.parcel-cache'));
+  await overlayFS.rimraf(path.join(__dirname, '/input/dist'));
+
+  if (typeof update === 'object') {
+    await update.setup();
+    update = update.update;
+  }
+
+  let b = await runBundle();
+
+  // update
+  await update(b);
+
+  // Run cached build
+  b = await runBundle();
+
+  return b;
+}
+
+async function assertThrows(fn) {
+  let threw = false;
+  let e;
+  try {
+    await fn();
+  } catch (err) {
+    threw = true;
+    e = err;
+  }
+
+  assert(threw, 'did not throw');
+  return e;
+}
+
+describe('cache', function() {
+  it('should support updating a JS file', async function() {
+    let b = await testCache(async b => {
+      assert.equal(await run(b), 4);
+      await overlayFS.writeFile(
+        path.join(__dirname, '/input/src/nested/test.js'),
+        'export default 4',
+      );
+    });
+
+    assert.equal(await run(b), 6);
+  });
+
+  it('should support adding a dependency', async function() {
+    let b = await testCache(async b => {
+      assert.equal(await run(b), 4);
+      await overlayFS.writeFile(
+        path.join(__dirname, '/input/src/nested/foo.js'),
+        'export default 6',
+      );
+      await overlayFS.writeFile(
+        path.join(__dirname, '/input/src/nested/test.js'),
+        'export {default} from "./foo";',
+      );
+    });
+
+    assert.equal(await run(b), 8);
+  });
+
+  it('should error when deleting a file', async function() {
+    let e = await assertThrows(async () => {
+      await testCache(async () => {
+        await overlayFS.unlink(
+          path.join(__dirname, '/input/src/nested/test.js'),
+        );
+      });
+    });
+
+    assert.equal(
+      e.message,
+      "Failed to resolve './nested/test' from './src/index.js'",
+    );
+  });
+
+  it('should error when starting parcel from a broken state with no changes', async function() {
+    await assertThrows(async () => {
+      await testCache(async () => {
+        await overlayFS.unlink(
+          path.join(__dirname, '/input/src/nested/test.js'),
+        );
+      });
+    });
+
+    // Do a third build from a failed state with no changes
+    let e = await assertThrows(async () => {
+      await runBundle();
+    });
+    assert.equal(
+      e.message,
+      "Failed to resolve './nested/test' from './src/index.js'",
+    );
+  });
+
+  describe('babel', function() {
```

Comment:
What do you think of the idea of moving plugin specific tests to the plugin packages? The advantages I see are that tests would be easier to find and easier to clean up if we migrate to a different plugin in the default config. It's not something I feel super strong about and it's definitely not a blocker, just something I've been thinking about.